crates/transpiler/src/passes/apply_layout.rs

```diff
@@ -0,0 +1,258 @@
+        let phys = layout_fn(virt);
+        let old_virt = ::std::mem::replace(&mut virtuals[phys.index()], virt);
+        assert_eq!(
+            old_virt,
+            VirtualQubit::MAX,
+            "both {virt:?} and {old_virt:?} were assigned to {phys:?}",
+        );
+        physicals.push(phys);
+    }
+    let unassigned = virtuals.iter_mut().enumerate().filter_map(|(i, virt)| {
+        (*virt == VirtualQubit::MAX).then_some((virt, PhysicalQubit::new(i as u32)))
+    });
+    for (ancilla, (slot, phys)) in (num_virtual_qubits..num_physical_qubits).zip(unassigned) {
+        physicals.push(phys);
+        *slot = VirtualQubit::new(ancilla);
+    }
+    let initial_layout = NLayout::from_vecs_unchecked(physicals, virtuals);
+
+    let final_permutation = cur_layout.output_permutation().map(|previous| {
+        let mut permutation = (0..num_physical_qubits as usize)
+            .map(Qubit::new)
+            .collect::<Vec<_>>();
+        for virt in (0..num_virtual_qubits).map(VirtualQubit::new) {
+            permutation[virt.to_phys(&initial_layout).index()] =
+                VirtualQubit::from(previous[virt.index()])
+                    .to_phys(&initial_layout)
+                    .into();
+        }
+        permutation
+    });
+    let mut virtual_qubits = dag.qubits().objects().to_vec();
+    let num_ancillas = num_physical_qubits - num_virtual_qubits;
+    if num_ancillas > 0 {
+        let reg_name = unique_ancilla_register_name(dag.qregs());
+        virtual_qubits.extend(QuantumRegister::new_owning(reg_name, num_ancillas).bits())
+    }
+
+    dag.make_physical(num_physical_qubits as usize);
+    dag.reindex_qargs(|virt| VirtualQubit::from(virt).to_phys(&initial_layout).into());
+    *cur_layout = TranspileLayout::new(
+        Some(initial_layout),
+        final_permutation,
+        virtual_qubits,
+        cur_layout.num_input_qubits(),
+    );
+}
+
+/// Permute the qubit indices of the [DAGCircuit] and the [TranspileLayout]
+///
+/// This is typically called to improve a previously set layout.
+///
+/// # Panics
+///
+/// * if the `layout_fn` returns a `PhysicalQubit` that is out of range.
+pub fn update_layout(
+    dag: &mut DAGCircuit,
+    cur_layout: &mut TranspileLayout,
+    mut layout_fn: impl FnMut(Qubit) -> Qubit,
+) {
+    dag.reindex_qargs(&mut layout_fn);
+    cur_layout.relabel_initial_layout(|q| layout_fn(q.into()).into());
+}
+
+fn unique_ancilla_register_name(qregs: &[QuantumRegister]) -> String {
+    let base = "ancilla";
+    if !qregs.iter().any(|qreg| qreg.name() == base) {
+        // Most likely we'll return out of here, and can avoid constructing a hash set.
+        return base.to_owned();
+    }
+    let names = qregs.iter().map(|qreg| qreg.name()).collect::<HashSet<_>>();
+    let mut i = 0;
+    loop {
+        let name = format!("{base}_{i}");
+        if !names.contains(name.as_str()) {
+            return name;
+        }
+        i += 1;
+    }
```

Comment:
This does change the exact register names from Python. While I agree this is a more foolproof implementation of what the intent was in `FullAncillaAllocation` in Python it just called the equivalent of:

```python
format!("ancilla_{}", QuantumRegister::anonymous_instance_count().fetch_add(1, Ordering::Relaxed)
```

While this will always use the lowest `i` value for `ancilla_{i}` that doesn't conflict with another name, it feels like it would have more overhead than just using the total number of created registers in the program like what python was doing. I guess the difference in a c context is that we may not be bumping the instance count for owned registers, was that your concern here?

Reply:
The Python-space version is bugged - there's nothing stopping you from defining `ancilla1` yourself and then `FullAncillaAllocation` will happily just attempt to re-use it and explode:
```python
from qiskit.circuit import QuantumCircuit, QuantumRegister
from qiskit.transpiler import CouplingMap, PassManager, passes

cm = CouplingMap.from_line(4)
qc = QuantumCircuit(QuantumRegister(1, "ancilla"), QuantumRegister(1, "ancilla0"))
PassManager([
    passes.TrivialLayout(cm),
    passes.FullAncillaAllocation(cm),
    passes.EnlargeWithAncilla(),
]).run(qc)
```
will raise a "duplicate register error".

Reply:
To speak to efficiency - I put the fast path in because 99.99% of the time, we're going to be going down that route, and it just outputs the regular name.  If not, we already have larger performance problems with having tonnes of registers (iirc, `QuantumCircuit.add_register` is quadratic in register count),

We can't even have a test asserting anything about this, because I didn't have to change the test suite.